{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Overview and references\n",
    "\n",
    "You will search several remote biological databases for gene sequences, annotations, and relationships between genes. You will then import the data you’ve found into a database that you will design.\n",
    "\n",
    "\n",
    "For this lab, you will import data of a particular type (e.g., a gene sequence, annotation, or pathway) into an appropriated-designed table and perform table joins that link together the data you’ve retrieved.\n",
    "\n",
    "\n",
    "You are expected to keep a thorough record of everything you did in your notebook. Any data that you’ve pulled from a web browser should be noted in this notebook. Copy and paste any terminal commands you used into a Markdown section and explain what the input was, what the tool did, and what the output was.\n",
    "\n",
    "\n",
    "### SQLite\n",
    "https://www.tutorialspoint.com/sqlite/index.htm https://docs.python.org/3/library/sqlite3.html\n",
    "\n",
    "### Relational Databases\n",
    "https://en.wikipedia.org/wiki/Relational_database https://mariadb.com/kb/en/library/relational-databases-basic-terms/ https://en.wikipedia.org/wiki/Associative_entity\n",
    "\n",
    "### BioPython\n",
    "https://biopython.org/DIST/docs/api/Bio.Entrez-module.html\n",
    "\n",
    "### Pandas\n",
    "https://pandas.pydata.org/pandas-docs/stable/index.html\n",
    "\n",
    "### Biological Databases\n",
    "https://genome.ucsc.edu/ https://www.ncbi.nlm.nih.gov/search/ https://reactome.org/ https://pfam.xfam.org/ https://www.genome.jp/kegg/\n",
    "\n",
    "# Background \n",
    "\n",
    "The lab you’re working in studies metabolism. Over the years, many students and post-docs have studied enzymes from organisms across the tree of life, from flies, worms, mice, and E. coli. Your PI wants you to build a database of these enzymes that future researchers in the lab can use going forward.\n",
    "\n",
    "\n",
    "# Database Design\n",
    "\n",
    "For this assignment, we’ll focus on three pathways: \n",
    "\n",
    "1. glycolysis\n",
    "2. the citric acid cycle\n",
    "3. the pentose phosphate pathway. \n",
    "\n",
    "We will consider genes in these pathways from \n",
    "\n",
    "1. Drosophila melanogaster\n",
    "2. Homo sapiens \n",
    "\n",
    "And we will construct the following tables:\n",
    "\n",
    "**Pathway Table**. Should include at least name and description. What fields should be included to enable joins with other tables?\n",
    "\n",
    "**Enzyme Table**. Should include at least name, function, and enzyme commission (EC) number. What fields should be included to enable joins with other tables?\n",
    "\n",
    "**Gene Table**. Should include at least name, description, organism, and nucleotide sequence. What fields should be included to enable joins with other tables?\n",
    "\n",
    "# Visualization\n",
    "Pandas provides a useful python interface for working with data tables. In this lab, we will use pandas DataFrames for visualization only. The rest of the lab should have been completed using SQL (and python interface to it). Read your SQL tables using the [`pandas.read_sql`](https://pandas.pydata.org/docs/reference/api/pandas.read_sql.html) method."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Part 1: Database/table creation\n",
    "\n",
    "\n",
    "Now that your tables are designed, try creating your database by using the function below. You might have to delete and re-create your database (the file) if mistakes are made and perform one final execution with the correct code.\n",
    "\n",
    "The command can be executed like this:\n",
    "\n",
    "```python\n",
    "command = \"\"\"\n",
    "    CREATE TABLE <fill in your code>\n",
    "\"\"\"\n",
    "execute_sql_command(command)\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%capture\n",
    "import sqlite3\n",
    "import pandas as pd\n",
    "!conda install -y -q -c conda-forge biopython"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def execute_sql_command(command: str, database_name: str = \"lab3.db\") -> None:\n",
    "    \"\"\"\n",
    "    Function to execute a SQL command from Python.\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    command: str\n",
    "        SQL command (use strings with three quotes on each side\n",
    "        so that it can be a multiline string\n",
    "    database_name: str\n",
    "        File name of the database (e.g, \"my.db\")\n",
    "    \n",
    "    Returns\n",
    "    -------\n",
    "    No return, executes the command\n",
    "    \"\"\"\n",
    "    \n",
    "    # will create if not present\n",
    "    conn = sqlite3.connect(database_name)\n",
    "    c = conn.cursor()\n",
    "    \n",
    "    c.execute(command)\n",
    "    # saves the changes\n",
    "    conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Part 1.1: Create pathways table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "command = \"\"\"\n",
    "    CREATE TABLE pathways\n",
    "    (name TEXT PRIMARY KEY,\n",
    "    description TEXT)\n",
    "    \n",
    "\"\"\"\n",
    "execute_sql_command(command)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Part 1.2: Create enzymes table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "command = \"\"\"\n",
    "    CREATE TABLE enzymes\n",
    "    (name TEXT,\n",
    "    description TEXT,\n",
    "    ecnum TEXT PRIMARY KEY,\n",
    "    pathway TEXT)\n",
    "    \n",
    "\"\"\"\n",
    "execute_sql_command(command)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Part 1.3: Create genes table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "command = \"\"\"\n",
    "    CREATE TABLE genes\n",
    "    (name TEXT,\n",
    "    description TEXT,\n",
    "    organism TEXT,\n",
    "    nuc_seq TEXT,\n",
    "    enzyme_ecnum TEXT) \n",
    "\"\"\"\n",
    "\n",
    "execute_sql_command(command)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Searching databases and compiling data\n",
    "\n",
    "A good place to start will be this map of glycolysis from KEGG: \n",
    "\n",
    "https://www.genome.jp/kegg-bin/show_pathway?map00010\n",
    "    \n",
    "There are a lot of genes in these pathways and how thoroughly you curate your database is up to you, but for the sake of time:\n",
    "\n",
    "1. Pick 4 enzymes from each of glycolysis, TCA cycle, and pentose phosphate. These enzymes should be relevant for both species. Look for \"change pathway type\" in the Kegg pathway map.\n",
    "2. Inspect information and sequences from Drosophila and human for those enzymes in KEGG database (no need to add anything in this notebook for this step, just inspect for yourself)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# INSERTing your data into your database\n",
    "\n",
    "- In your pathways table, you should have a row per pathway (3 in total).\n",
    "- In your enzymes table, you should have one row per enzyme (12 in total).\n",
    "- In your gene table, you’ll be creating `4 enzymes * 3 pathways * 2 organisms = 24 rows`."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Part 2: populate pathways table\n",
    "For each pathway, insert a row into the database using the `INSERT` operation. Once you are done, print the contents of the table using `SELECT` and `pd.read_sql`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Source: https://www.genome.jp/pathway/map00010\n",
    "insert1 = \"\"\"\n",
    "    INSERT INTO pathways\n",
    "    VALUES\n",
    "    ('glycolysis',\n",
    "    'Glycolysis is the process of converting glucose into pyruvate and generating small amounts of ATP (energy) and NADH (reducing power).')\n",
    "\"\"\"\n",
    "\n",
    "#Source: https://www.genome.jp/pathway/map00020\n",
    "insert2 = \"\"\"\n",
    "    INSERT INTO pathways\n",
    "    VALUES\n",
    "    ('citric acid cycle',\n",
    "    'The citrate cycle (TCA cycle, Krebs cycle) is an important aerobic pathway for the final steps of the oxidation of carbohydrates and fatty acids.')\n",
    "\"\"\"\n",
    "\n",
    "#Source: https://www.genome.jp/pathway/map00030\n",
    "insert3 = \"\"\"\n",
    "    INSERT INTO pathways\n",
    "    VALUES\n",
    "    ('pentose phosphate pathway',\n",
    "    'The pentose phosphate pathway is a process of glucose turnover that produces NADPH as reducing equivalents and pentoses as essential parts of nucleotides.')\n",
    "\"\"\"\n",
    "\n",
    "commands = [insert1, insert2, insert3]\n",
    "for command in commands: execute_sql_command(command)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>glycolysis</td>\n",
       "      <td>Glycolysis is the process of converting glucos...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>citric acid cycle</td>\n",
       "      <td>The citrate cycle (TCA cycle, Krebs cycle) is ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>pentose phosphate pathway</td>\n",
       "      <td>The pentose phosphate pathway is a process of ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        name  \\\n",
       "0                 glycolysis   \n",
       "1          citric acid cycle   \n",
       "2  pentose phosphate pathway   \n",
       "\n",
       "                                         description  \n",
       "0  Glycolysis is the process of converting glucos...  \n",
       "1  The citrate cycle (TCA cycle, Krebs cycle) is ...  \n",
       "2  The pentose phosphate pathway is a process of ...  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "conn = sqlite3.connect('lab3.db')\n",
    "pd.read_sql('SELECT * FROM pathways', conn)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Part 3: populate enzymes table\n",
    "Similarly, populate the enzymes table and print its contents."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "#SOURCE: https://www.genome.jp/entry/3.1.3.9\n",
    "insert1 = \"\"\"\n",
    "    INSERT INTO enzymes\n",
    "    VALUES\n",
    "    ('glucose-6-phosphatase',\n",
    "    'Wide distribution in animal tissues.',\n",
    "    '3.1.3.9',\n",
    "    'glycolysis')\n",
    "\"\"\"\n",
    "\n",
    "#SOURCE: https://www.genome.jp/entry/5.4.2.2\n",
    "insert2 = \"\"\"\n",
    "    INSERT INTO enzymes\n",
    "    VALUES\n",
    "    ('phosphoglucomutase',\n",
    "    'Maximum activity is only obtained in the presence of alpha-D-glucose 1,6-bisphosphate.',\n",
    "    '5.4.2.2',\n",
    "    'glycolysis')\n",
    "\"\"\"\n",
    "\n",
    "#SOURCE: https://www.genome.jp/entry/2.7.1.1\n",
    "insert3 = \"\"\"\n",
    "    INSERT INTO enzymes\n",
    "    VALUES\n",
    "    ('hexokinase',\n",
    "    'D-Glucose, D-mannose, D-fructose, sorbitol and D-glucosamine can act as acceptors; ITP and dATP can act as donors.',\n",
    "    '2.7.1.1',\n",
    "    'glycolysis')\n",
    "\"\"\"\n",
    "\n",
    "#SOURCE: https://www.genome.jp/entry/5.1.3.3\n",
    "insert4 = \"\"\"\n",
    "    INSERT INTO enzymes\n",
    "    VALUES\n",
    "    ('aldose 1-epimerase',\n",
    "    'This enzyme catalyses the first step in galactose metabolism.',\n",
    "    '5.1.3.3',\n",
    "    'glycolysis')\n",
    "\"\"\"\n",
    "\n",
    "#SOURCE: https://www.genome.jp/entry/4.1.1.32\n",
    "insert5 = \"\"\"\n",
    "    INSERT INTO enzymes\n",
    "    VALUES\n",
    "    ('phosphoenolpyruvate carboxykinase (GTP)',\n",
    "    'ITP can act as phosphate donor.',\n",
    "    '4.1.1.32',\n",
    "    'citric acid cycle')\n",
    "\"\"\"\n",
    "\n",
    "#SOURCE: https://www.genome.jp/entry/2.3.1.12\n",
    "insert6 = \"\"\"\n",
    "    INSERT INTO enzymes\n",
    "    VALUES\n",
    "    ('dihydrolipoyllysine-residue acetyltransferase',\n",
    "    'Belongs to class: Transferring groups other than aminoacyl groups',\n",
    "    '2.3.1.12',\n",
    "    'citric acid cycle')\n",
    "\"\"\"\n",
    "\n",
    "#SOURCE: https://www.genome.jp/entry/1.2.4.1\n",
    "insert7 = \"\"\"\n",
    "    INSERT INTO enzymes\n",
    "    VALUES\n",
    "    ('pyruvate dehydrogenase (acetyl-transferring)',\n",
    "    'Contains thiamine diphosphate.',\n",
    "    '1.2.4.1',\n",
    "    'citric acid cycle')\n",
    "\"\"\"\n",
    "\n",
    "#SOURCE: https://www.genome.jp/entry/4.2.1.3\n",
    "insert8 = \"\"\"\n",
    "    INSERT INTO enzymes\n",
    "    VALUES\n",
    "    ('aconitate hydratase',\n",
    "    'Besides interconverting citrate and cis-aconitate, it also interconverts cis-aconitate with isocitrate and, hence, interconverts citrate and isocitrate.',\n",
    "    '4.2.1.3',\n",
    "    'citric acid cycle')\n",
    "\"\"\"\n",
    "\n",
    "#SOURCE: https://www.genome.jp/entry/1.1.1.49\n",
    "insert9 = \"\"\"\n",
    "    INSERT INTO enzymes\n",
    "    VALUES\n",
    "    ('glucose-6-phosphate dehydrogenase (NADP+)',\n",
    "    'The enzyme catalyses a step of the pentose phosphate pathway.',\n",
    "    '1.1.1.49',\n",
    "    'pentose phosphate pathway')\n",
    "\"\"\"\n",
    "\n",
    "#SOURCE: https://www.genome.jp/entry/5.3.1.9\n",
    "insert10 = \"\"\"\n",
    "    INSERT INTO enzymes\n",
    "    VALUES\n",
    "    ('glucose-6-phosphate isomerase',\n",
    "    'Also catalyses the anomerization of D-glucose 6-phosphate.',\n",
    "    '5.3.1.9',\n",
    "    'pentose phosphate pathway')\n",
    "\"\"\"\n",
    "\n",
    "#SOURCE: https://www.genome.jp/entry/3.1.1.31\n",
    "insert11 = \"\"\"\n",
    "    INSERT INTO enzymes\n",
    "    VALUES\n",
    "    ('6-phosphogluconolactonase',\n",
    "    'Belongs to the class: Carboxylic-ester hydrolases',\n",
    "    '3.1.1.31',\n",
    "    'pentose phosphate pathway')\n",
    "\"\"\"\n",
    "\n",
    "#SOURCE: https://www.genome.jp/entry/5.1.3.1\n",
    "insert12 = \"\"\"\n",
    "    INSERT INTO enzymes\n",
    "    VALUES\n",
    "    ('ribulose-phosphate 3-epimerase',\n",
    "    'The enzyme also converts D-erythrose 4-phosphate into D-erythrulose 4-phosphate and D-threose 4-phosphate.',\n",
    "    '5.1.3.1',\n",
    "    'pentose phosphate pathway')\n",
    "\"\"\"\n",
    "\n",
    "commands = [insert1, insert2, insert3, insert4, insert5, insert6, insert7, insert8, insert9, insert10, insert11, insert12]\n",
    "for command in commands: execute_sql_command(command)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>description</th>\n",
       "      <th>ecnum</th>\n",
       "      <th>pathway</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>glucose-6-phosphatase</td>\n",
       "      <td>Wide distribution in animal tissues.</td>\n",
       "      <td>3.1.3.9</td>\n",
       "      <td>glycolysis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>phosphoglucomutase</td>\n",
       "      <td>Maximum activity is only obtained in the prese...</td>\n",
       "      <td>5.4.2.2</td>\n",
       "      <td>glycolysis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>hexokinase</td>\n",
       "      <td>D-Glucose, D-mannose, D-fructose, sorbitol and...</td>\n",
       "      <td>2.7.1.1</td>\n",
       "      <td>glycolysis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>aldose 1-epimerase</td>\n",
       "      <td>This enzyme catalyses the first step in galact...</td>\n",
       "      <td>5.1.3.3</td>\n",
       "      <td>glycolysis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>phosphoenolpyruvate carboxykinase (GTP)</td>\n",
       "      <td>ITP can act as phosphate donor.</td>\n",
       "      <td>4.1.1.32</td>\n",
       "      <td>citric acid cycle</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>dihydrolipoyllysine-residue acetyltransferase</td>\n",
       "      <td>Belongs to class: Transferring groups other th...</td>\n",
       "      <td>2.3.1.12</td>\n",
       "      <td>citric acid cycle</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>pyruvate dehydrogenase (acetyl-transferring)</td>\n",
       "      <td>Contains thiamine diphosphate.</td>\n",
       "      <td>1.2.4.1</td>\n",
       "      <td>citric acid cycle</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>aconitate hydratase</td>\n",
       "      <td>Besides interconverting citrate and cis-aconit...</td>\n",
       "      <td>4.2.1.3</td>\n",
       "      <td>citric acid cycle</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>glucose-6-phosphate dehydrogenase (NADP+)</td>\n",
       "      <td>The enzyme catalyses a step of the pentose pho...</td>\n",
       "      <td>1.1.1.49</td>\n",
       "      <td>pentose phosphate pathway</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>glucose-6-phosphate isomerase</td>\n",
       "      <td>Also catalyses the anomerization of D-glucose ...</td>\n",
       "      <td>5.3.1.9</td>\n",
       "      <td>pentose phosphate pathway</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>6-phosphogluconolactonase</td>\n",
       "      <td>Belongs to the class: Carboxylic-ester hydrolases</td>\n",
       "      <td>3.1.1.31</td>\n",
       "      <td>pentose phosphate pathway</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ribulose-phosphate 3-epimerase</td>\n",
       "      <td>The enzyme also converts D-erythrose 4-phospha...</td>\n",
       "      <td>5.1.3.1</td>\n",
       "      <td>pentose phosphate pathway</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             name  \\\n",
       "0                           glucose-6-phosphatase   \n",
       "1                              phosphoglucomutase   \n",
       "2                                      hexokinase   \n",
       "3                              aldose 1-epimerase   \n",
       "4         phosphoenolpyruvate carboxykinase (GTP)   \n",
       "5   dihydrolipoyllysine-residue acetyltransferase   \n",
       "6    pyruvate dehydrogenase (acetyl-transferring)   \n",
       "7                             aconitate hydratase   \n",
       "8       glucose-6-phosphate dehydrogenase (NADP+)   \n",
       "9                   glucose-6-phosphate isomerase   \n",
       "10                      6-phosphogluconolactonase   \n",
       "11                 ribulose-phosphate 3-epimerase   \n",
       "\n",
       "                                          description     ecnum  \\\n",
       "0                Wide distribution in animal tissues.   3.1.3.9   \n",
       "1   Maximum activity is only obtained in the prese...   5.4.2.2   \n",
       "2   D-Glucose, D-mannose, D-fructose, sorbitol and...   2.7.1.1   \n",
       "3   This enzyme catalyses the first step in galact...   5.1.3.3   \n",
       "4                     ITP can act as phosphate donor.  4.1.1.32   \n",
       "5   Belongs to class: Transferring groups other th...  2.3.1.12   \n",
       "6                      Contains thiamine diphosphate.   1.2.4.1   \n",
       "7   Besides interconverting citrate and cis-aconit...   4.2.1.3   \n",
       "8   The enzyme catalyses a step of the pentose pho...  1.1.1.49   \n",
       "9   Also catalyses the anomerization of D-glucose ...   5.3.1.9   \n",
       "10  Belongs to the class: Carboxylic-ester hydrolases  3.1.1.31   \n",
       "11  The enzyme also converts D-erythrose 4-phospha...   5.1.3.1   \n",
       "\n",
       "                      pathway  \n",
       "0                  glycolysis  \n",
       "1                  glycolysis  \n",
       "2                  glycolysis  \n",
       "3                  glycolysis  \n",
       "4           citric acid cycle  \n",
       "5           citric acid cycle  \n",
       "6           citric acid cycle  \n",
       "7           citric acid cycle  \n",
       "8   pentose phosphate pathway  \n",
       "9   pentose phosphate pathway  \n",
       "10  pentose phosphate pathway  \n",
       "11  pentose phosphate pathway  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "conn = sqlite3.connect('lab3.db')\n",
    "pd.read_sql('SELECT * FROM enzymes', conn)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Part 4: retrieve gene information\n",
    "\n",
    "For genes, you will now automate this process using `Bio.Entrez`, which can be used to pull entire annotations for each enzyme from each organism—and more. You will ask it to return data as `GenBank` and later will use `Bio.SeqIO` to read the written `GenBank` file.\n",
    "\n",
    "For this part, complete the commented code in the function below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import Entrez\n",
    "from Bio import SeqIO\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.Seq import Seq\n",
    "\n",
    "Entrez.email = 'travishanly@berkeley.edu'\n",
    "\n",
    "def save_enzyme(ec_number: str, organism: str, file: str) -> None:\n",
    "    \"\"\"\n",
    "    Pull and save most relevant annotation for enzyme/organism pair\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    ec_number: str\n",
    "        Number assigned by the Enzyme Commission to designate a particular enzyme \n",
    "        or by the Chemical Abstracts Service (CAS) for Registry Numbers. \n",
    "    organism: str\n",
    "        Organism name\n",
    "    file: str\n",
    "        Name of saved file. SHOULD HAVE .gb extension\n",
    "        Points will be deducted if these files are committed to the repo.\n",
    "    \n",
    "    Returns\n",
    "    -------\n",
    "    No return, executes the command\n",
    "    \"\"\"\n",
    "    term = ec_number  + \" AND \" + organism + \"[Organism] AND 1:10000[SLEN]\"\n",
    "    handle = Entrez.esearch(db = 'nucleotide',\n",
    "                           term = term ,\n",
    "                           sort = 'relevance',\n",
    "                           idtype = 'acc')\n",
    "    for i in Entrez.read(handle)['IdList']:\n",
    "        handle = Entrez.efetch(db = 'nucleotide', id = i, rettype = 'gb', retmode = 'text', retmax=1)\n",
    "        seq = SeqIO.read(handle, 'genbank')\n",
    "        if seq.annotations['organism'] == organism:\n",
    "            # write the sequence to file using `SeqIO.write` in genbank (.gb) format.\n",
    "            SeqIO.write(seq, file, \"genbank\")\n",
    "            return"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Run `save_enzyme` 24 times (`4 enzymes * 3 pathways * 2 organisms = 24`). This can take a while."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "conn = sqlite3.connect('lab3.db')\n",
    "enzyme_numbers = list(pd.read_sql('SELECT ecnum FROM enzymes', conn)['ecnum'])\n",
    "organisms = [\"Homo sapiens\", \"Drosophila melanogaster\"]\n",
    "\n",
    "for organism in organisms:\n",
    "    for ecnum in enzyme_numbers:\n",
    "        file = organism + \"_\" + ecnum.replace(\".\", \"-\") + \".gb\"\n",
    "        save_enzyme(ecnum, organism, file)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Part 5: populate genes table\n",
    "\n",
    "Since we managed to pull the data directly into `genbank` files using `Bio.Entrez`, a for loop over the files can be used to insert your data into your table (make sure to use `execute_sql_command`). Print the table contents."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "enzyme_numbers = list(pd.read_sql('SELECT ecnum FROM enzymes', conn)['ecnum'])\n",
    "organisms = [\"Homo sapiens\", \"Drosophila melanogaster\"]\n",
    "\n",
    "for organism in organisms:\n",
    "    for ecnum in enzyme_numbers:\n",
    "        file = organism + \"_\" + ecnum.replace(\".\", \"-\") + \".gb\"\n",
    "        seq = SeqIO.read(file, \"genbank\")\n",
    "        command = f\"\"\"INSERT INTO genes\n",
    "                VALUES\n",
    "                ('{seq.id}',\n",
    "                '{seq.description.replace(\"'\", \"\")}',\n",
    "                '{organism}',\n",
    "                '{seq.seq}',\n",
    "                '{ecnum}')\"\"\"\n",
    "        execute_sql_command(command)       "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>description</th>\n",
       "      <th>organism</th>\n",
       "      <th>nuc_seq</th>\n",
       "      <th>enzyme_ecnum</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AK302341.1</td>\n",
       "      <td>Homo sapiens cDNA FLJ56794 complete cds, highl...</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>AAAACGGTCGTACACTTCGGGGCTGCGAGCGCGGAGGGCGACGACG...</td>\n",
       "      <td>1.1.1.49</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AK296457.1</td>\n",
       "      <td>Homo sapiens cDNA FLJ52314 complete cds, highl...</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>ACCCGCGTCGTGCCTCCTGGGTTGTGAGGAGTCGCCGCTGCCGCCA...</td>\n",
       "      <td>1.2.4.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AK299562.1</td>\n",
       "      <td>Homo sapiens cDNA FLJ51063 complete cds, highl...</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>CCCGTTCGTCGCAACAGCGTGACTACAGGGTATGGCGGGGTCCGGG...</td>\n",
       "      <td>2.3.1.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AK294428.1</td>\n",
       "      <td>Homo sapiens cDNA FLJ56506 complete cds, highl...</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>GAGGAGCCGCCGAGCAGCCGCCGGAGGACCACGGCTCGCCAGGGCT...</td>\n",
       "      <td>2.7.1.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CD580465.2</td>\n",
       "      <td>EST_PSF016769 hEx1 (RZPD no. 800) Homo sapiens...</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>CTCCCCGCCGCCGCCCTCGCCATGGCCGCGCCGGCCCCGGGCCTCA...</td>\n",
       "      <td>3.1.1.31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>AK303771.1</td>\n",
       "      <td>Homo sapiens cDNA FLJ51443 complete cds, highl...</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>ATAGCAGAGCAATCACCACCAAGCCTGGAATAACTGCAAGGGCTCT...</td>\n",
       "      <td>3.1.3.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>AK316206.1</td>\n",
       "      <td>Homo sapiens cDNA, FLJ79105 complete cds, high...</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>GTGGCTCGCTTCGCCGCGCTCCCTCCTTCCCCGCCTTCCATACCTC...</td>\n",
       "      <td>4.1.1.32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>AK302691.1</td>\n",
       "      <td>Homo sapiens cDNA FLJ51705 complete cds, highl...</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>ATGCGACCTCATCTTTGTCAGTGCACAAAATGGCGCCCTACAGCCT...</td>\n",
       "      <td>4.2.1.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>AK303184.1</td>\n",
       "      <td>Homo sapiens cDNA FLJ52962 complete cds, highl...</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>AACTTGCTTTTGGGAGCCAGCGGTATGGCGTCGGGCTGCAAGATTG...</td>\n",
       "      <td>5.1.3.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>NM_138801.3</td>\n",
       "      <td>Homo sapiens galactose mutarotase (GALM), mRNA</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>GCTGGAGTTTGAAGAGCGGGCAGTGGCTGCACACGCCAAACTTTCC...</td>\n",
       "      <td>5.1.3.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>AK294396.1</td>\n",
       "      <td>Homo sapiens cDNA FLJ56429 complete cds, highl...</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>ACACATATGTATCCACAGCGCCTAGAACACTGCCTGTAACGTGGAA...</td>\n",
       "      <td>5.3.1.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>AK300100.1</td>\n",
       "      <td>Homo sapiens cDNA FLJ50663 complete cds, highl...</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>GACCTGCTTTGTGTGTGGGTGCGAGTGGGTGTGTGGCCCTTAACTT...</td>\n",
       "      <td>5.4.2.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>NM_001298527.1</td>\n",
       "      <td>Drosophila melanogaster zwischenferment, trans...</td>\n",
       "      <td>Drosophila melanogaster</td>\n",
       "      <td>AGTTTTGACAGTTGAAAAAAGTTACTTAGTGAAAAAAACAAGTAGA...</td>\n",
       "      <td>1.1.1.49</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>NM_167019.2</td>\n",
       "      <td>Drosophila melanogaster pyruvate dehydrogenase...</td>\n",
       "      <td>Drosophila melanogaster</td>\n",
       "      <td>CAAACGGTTGTTATAGTTGTGCAATATAGTAGTTGATAAACTAATC...</td>\n",
       "      <td>1.2.4.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>NM_001273267.1</td>\n",
       "      <td>Drosophila melanogaster midline uncoordinated ...</td>\n",
       "      <td>Drosophila melanogaster</td>\n",
       "      <td>CGTTCCGTCTGCGTTCACACTGACCTCCGTAAGGTTATCGGCATAA...</td>\n",
       "      <td>2.3.1.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>NM_079935.4</td>\n",
       "      <td>Drosophila melanogaster hexokinase C (Hex-C), ...</td>\n",
       "      <td>Drosophila melanogaster</td>\n",
       "      <td>CCTGGTTAGTCGTTACTCATCGTTCGAGGCGGTTACAACTAAGACA...</td>\n",
       "      <td>2.7.1.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>NM_132428.3</td>\n",
       "      <td>Drosophila melanogaster uncharacterized protei...</td>\n",
       "      <td>Drosophila melanogaster</td>\n",
       "      <td>GAGGGAATGGTGATTACTGACAAATCGAAGTCGAGTCGAGTTTTAT...</td>\n",
       "      <td>3.1.1.31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>NM_001103593.2</td>\n",
       "      <td>Drosophila melanogaster Glucose-6-Phosphatase ...</td>\n",
       "      <td>Drosophila melanogaster</td>\n",
       "      <td>ATTTCATATTGGCAGAAAGACTTGAAAATATAAAAGGAGCCAGTTG...</td>\n",
       "      <td>3.1.3.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>AA567790.1</td>\n",
       "      <td>HL01774.5prime HL Drosophila melanogaster head...</td>\n",
       "      <td>Drosophila melanogaster</td>\n",
       "      <td>AAATCGGGCACAAGGAAAGATCAAGTGAAAAGCCTACTTTCAGTGG...</td>\n",
       "      <td>4.1.1.32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>NM_001299190.1</td>\n",
       "      <td>Drosophila melanogaster mitochondrial aconitas...</td>\n",
       "      <td>Drosophila melanogaster</td>\n",
       "      <td>AATTTTGGCGGGAAACAAGCCGGCGGAACATTGCTGACGCGTATGC...</td>\n",
       "      <td>4.2.1.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>NM_001273851.1</td>\n",
       "      <td>Drosophila melanogaster Ribulose-5-phosphate-3...</td>\n",
       "      <td>Drosophila melanogaster</td>\n",
       "      <td>AGCTCCCCTCTGTGTGACCTTAAGTCAACAAGACGCCACCCGCGAA...</td>\n",
       "      <td>5.1.3.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>NM_135670.2</td>\n",
       "      <td>Drosophila melanogaster uncharacterized protei...</td>\n",
       "      <td>Drosophila melanogaster</td>\n",
       "      <td>ATTTTAAATCAAAATAACCTGGCATTTTCTATCAATTTTTAATGAG...</td>\n",
       "      <td>5.1.3.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>NM_165636.2</td>\n",
       "      <td>Drosophila melanogaster phosphoglucose isomera...</td>\n",
       "      <td>Drosophila melanogaster</td>\n",
       "      <td>ATTTGACATTGAAAGGTGCTCGCGAATTCCAGAAAGGACGTTGGAA...</td>\n",
       "      <td>5.3.1.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>NM_079936.3</td>\n",
       "      <td>Drosophila melanogaster phosphoglucose mutase ...</td>\n",
       "      <td>Drosophila melanogaster</td>\n",
       "      <td>TCCAGTTGGAATTCAGCCAGCAGCCGGAAAACTCCAGTGACATGTC...</td>\n",
       "      <td>5.4.2.2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              name                                        description  \\\n",
       "0       AK302341.1  Homo sapiens cDNA FLJ56794 complete cds, highl...   \n",
       "1       AK296457.1  Homo sapiens cDNA FLJ52314 complete cds, highl...   \n",
       "2       AK299562.1  Homo sapiens cDNA FLJ51063 complete cds, highl...   \n",
       "3       AK294428.1  Homo sapiens cDNA FLJ56506 complete cds, highl...   \n",
       "4       CD580465.2  EST_PSF016769 hEx1 (RZPD no. 800) Homo sapiens...   \n",
       "5       AK303771.1  Homo sapiens cDNA FLJ51443 complete cds, highl...   \n",
       "6       AK316206.1  Homo sapiens cDNA, FLJ79105 complete cds, high...   \n",
       "7       AK302691.1  Homo sapiens cDNA FLJ51705 complete cds, highl...   \n",
       "8       AK303184.1  Homo sapiens cDNA FLJ52962 complete cds, highl...   \n",
       "9      NM_138801.3     Homo sapiens galactose mutarotase (GALM), mRNA   \n",
       "10      AK294396.1  Homo sapiens cDNA FLJ56429 complete cds, highl...   \n",
       "11      AK300100.1  Homo sapiens cDNA FLJ50663 complete cds, highl...   \n",
       "12  NM_001298527.1  Drosophila melanogaster zwischenferment, trans...   \n",
       "13     NM_167019.2  Drosophila melanogaster pyruvate dehydrogenase...   \n",
       "14  NM_001273267.1  Drosophila melanogaster midline uncoordinated ...   \n",
       "15     NM_079935.4  Drosophila melanogaster hexokinase C (Hex-C), ...   \n",
       "16     NM_132428.3  Drosophila melanogaster uncharacterized protei...   \n",
       "17  NM_001103593.2  Drosophila melanogaster Glucose-6-Phosphatase ...   \n",
       "18      AA567790.1  HL01774.5prime HL Drosophila melanogaster head...   \n",
       "19  NM_001299190.1  Drosophila melanogaster mitochondrial aconitas...   \n",
       "20  NM_001273851.1  Drosophila melanogaster Ribulose-5-phosphate-3...   \n",
       "21     NM_135670.2  Drosophila melanogaster uncharacterized protei...   \n",
       "22     NM_165636.2  Drosophila melanogaster phosphoglucose isomera...   \n",
       "23     NM_079936.3  Drosophila melanogaster phosphoglucose mutase ...   \n",
       "\n",
       "                   organism  \\\n",
       "0              Homo sapiens   \n",
       "1              Homo sapiens   \n",
       "2              Homo sapiens   \n",
       "3              Homo sapiens   \n",
       "4              Homo sapiens   \n",
       "5              Homo sapiens   \n",
       "6              Homo sapiens   \n",
       "7              Homo sapiens   \n",
       "8              Homo sapiens   \n",
       "9              Homo sapiens   \n",
       "10             Homo sapiens   \n",
       "11             Homo sapiens   \n",
       "12  Drosophila melanogaster   \n",
       "13  Drosophila melanogaster   \n",
       "14  Drosophila melanogaster   \n",
       "15  Drosophila melanogaster   \n",
       "16  Drosophila melanogaster   \n",
       "17  Drosophila melanogaster   \n",
       "18  Drosophila melanogaster   \n",
       "19  Drosophila melanogaster   \n",
       "20  Drosophila melanogaster   \n",
       "21  Drosophila melanogaster   \n",
       "22  Drosophila melanogaster   \n",
       "23  Drosophila melanogaster   \n",
       "\n",
       "                                              nuc_seq enzyme_ecnum  \n",
       "0   AAAACGGTCGTACACTTCGGGGCTGCGAGCGCGGAGGGCGACGACG...     1.1.1.49  \n",
       "1   ACCCGCGTCGTGCCTCCTGGGTTGTGAGGAGTCGCCGCTGCCGCCA...      1.2.4.1  \n",
       "2   CCCGTTCGTCGCAACAGCGTGACTACAGGGTATGGCGGGGTCCGGG...     2.3.1.12  \n",
       "3   GAGGAGCCGCCGAGCAGCCGCCGGAGGACCACGGCTCGCCAGGGCT...      2.7.1.1  \n",
       "4   CTCCCCGCCGCCGCCCTCGCCATGGCCGCGCCGGCCCCGGGCCTCA...     3.1.1.31  \n",
       "5   ATAGCAGAGCAATCACCACCAAGCCTGGAATAACTGCAAGGGCTCT...      3.1.3.9  \n",
       "6   GTGGCTCGCTTCGCCGCGCTCCCTCCTTCCCCGCCTTCCATACCTC...     4.1.1.32  \n",
       "7   ATGCGACCTCATCTTTGTCAGTGCACAAAATGGCGCCCTACAGCCT...      4.2.1.3  \n",
       "8   AACTTGCTTTTGGGAGCCAGCGGTATGGCGTCGGGCTGCAAGATTG...      5.1.3.1  \n",
       "9   GCTGGAGTTTGAAGAGCGGGCAGTGGCTGCACACGCCAAACTTTCC...      5.1.3.3  \n",
       "10  ACACATATGTATCCACAGCGCCTAGAACACTGCCTGTAACGTGGAA...      5.3.1.9  \n",
       "11  GACCTGCTTTGTGTGTGGGTGCGAGTGGGTGTGTGGCCCTTAACTT...      5.4.2.2  \n",
       "12  AGTTTTGACAGTTGAAAAAAGTTACTTAGTGAAAAAAACAAGTAGA...     1.1.1.49  \n",
       "13  CAAACGGTTGTTATAGTTGTGCAATATAGTAGTTGATAAACTAATC...      1.2.4.1  \n",
       "14  CGTTCCGTCTGCGTTCACACTGACCTCCGTAAGGTTATCGGCATAA...     2.3.1.12  \n",
       "15  CCTGGTTAGTCGTTACTCATCGTTCGAGGCGGTTACAACTAAGACA...      2.7.1.1  \n",
       "16  GAGGGAATGGTGATTACTGACAAATCGAAGTCGAGTCGAGTTTTAT...     3.1.1.31  \n",
       "17  ATTTCATATTGGCAGAAAGACTTGAAAATATAAAAGGAGCCAGTTG...      3.1.3.9  \n",
       "18  AAATCGGGCACAAGGAAAGATCAAGTGAAAAGCCTACTTTCAGTGG...     4.1.1.32  \n",
       "19  AATTTTGGCGGGAAACAAGCCGGCGGAACATTGCTGACGCGTATGC...      4.2.1.3  \n",
       "20  AGCTCCCCTCTGTGTGACCTTAAGTCAACAAGACGCCACCCGCGAA...      5.1.3.1  \n",
       "21  ATTTTAAATCAAAATAACCTGGCATTTTCTATCAATTTTTAATGAG...      5.1.3.3  \n",
       "22  ATTTGACATTGAAAGGTGCTCGCGAATTCCAGAAAGGACGTTGGAA...      5.3.1.9  \n",
       "23  TCCAGTTGGAATTCAGCCAGCAGCCGGAAAACTCCAGTGACATGTC...      5.4.2.2  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "conn = sqlite3.connect('lab3.db')\n",
    "pd.read_sql('SELECT * FROM genes', conn)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Part 6: Query data with JOIN\n",
    "\n",
    "You should have added fields to your tables to link data together. For this part of the lab we would like to retrieve information from the three tables and print it with `pd.read_sql`."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Part 6.1\n",
    "Make a join of the three tables and display the results."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>gene_description</th>\n",
       "      <th>organism</th>\n",
       "      <th>nuc_seq</th>\n",
       "      <th>enzyme</th>\n",
       "      <th>enzyme_description</th>\n",
       "      <th>ecnum</th>\n",
       "      <th>pathway</th>\n",
       "      <th>pathway_description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AK302341.1</td>\n",
       "      <td>Homo sapiens cDNA FLJ56794 complete cds, highl...</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>AAAACGGTCGTACACTTCGGGGCTGCGAGCGCGGAGGGCGACGACG...</td>\n",
       "      <td>glucose-6-phosphate dehydrogenase (NADP+)</td>\n",
       "      <td>The enzyme catalyses a step of the pentose pho...</td>\n",
       "      <td>1.1.1.49</td>\n",
       "      <td>pentose phosphate pathway</td>\n",
       "      <td>The pentose phosphate pathway is a process of ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AK296457.1</td>\n",
       "      <td>Homo sapiens cDNA FLJ52314 complete cds, highl...</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>ACCCGCGTCGTGCCTCCTGGGTTGTGAGGAGTCGCCGCTGCCGCCA...</td>\n",
       "      <td>pyruvate dehydrogenase (acetyl-transferring)</td>\n",
       "      <td>Contains thiamine diphosphate.</td>\n",
       "      <td>1.2.4.1</td>\n",
       "      <td>citric acid cycle</td>\n",
       "      <td>The citrate cycle (TCA cycle, Krebs cycle) is ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AK299562.1</td>\n",
       "      <td>Homo sapiens cDNA FLJ51063 complete cds, highl...</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>CCCGTTCGTCGCAACAGCGTGACTACAGGGTATGGCGGGGTCCGGG...</td>\n",
       "      <td>dihydrolipoyllysine-residue acetyltransferase</td>\n",
       "      <td>Belongs to class: Transferring groups other th...</td>\n",
       "      <td>2.3.1.12</td>\n",
       "      <td>citric acid cycle</td>\n",
       "      <td>The citrate cycle (TCA cycle, Krebs cycle) is ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AK294428.1</td>\n",
       "      <td>Homo sapiens cDNA FLJ56506 complete cds, highl...</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>GAGGAGCCGCCGAGCAGCCGCCGGAGGACCACGGCTCGCCAGGGCT...</td>\n",
       "      <td>hexokinase</td>\n",
       "      <td>D-Glucose, D-mannose, D-fructose, sorbitol and...</td>\n",
       "      <td>2.7.1.1</td>\n",
       "      <td>glycolysis</td>\n",
       "      <td>Glycolysis is the process of converting glucos...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CD580465.2</td>\n",
       "      <td>EST_PSF016769 hEx1 (RZPD no. 800) Homo sapiens...</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>CTCCCCGCCGCCGCCCTCGCCATGGCCGCGCCGGCCCCGGGCCTCA...</td>\n",
       "      <td>6-phosphogluconolactonase</td>\n",
       "      <td>Belongs to the class: Carboxylic-ester hydrolases</td>\n",
       "      <td>3.1.1.31</td>\n",
       "      <td>pentose phosphate pathway</td>\n",
       "      <td>The pentose phosphate pathway is a process of ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>AK303771.1</td>\n",
       "      <td>Homo sapiens cDNA FLJ51443 complete cds, highl...</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>ATAGCAGAGCAATCACCACCAAGCCTGGAATAACTGCAAGGGCTCT...</td>\n",
       "      <td>glucose-6-phosphatase</td>\n",
       "      <td>Wide distribution in animal tissues.</td>\n",
       "      <td>3.1.3.9</td>\n",
       "      <td>glycolysis</td>\n",
       "      <td>Glycolysis is the process of converting glucos...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>AK316206.1</td>\n",
       "      <td>Homo sapiens cDNA, FLJ79105 complete cds, high...</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>GTGGCTCGCTTCGCCGCGCTCCCTCCTTCCCCGCCTTCCATACCTC...</td>\n",
       "      <td>phosphoenolpyruvate carboxykinase (GTP)</td>\n",
       "      <td>ITP can act as phosphate donor.</td>\n",
       "      <td>4.1.1.32</td>\n",
       "      <td>citric acid cycle</td>\n",
       "      <td>The citrate cycle (TCA cycle, Krebs cycle) is ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>AK302691.1</td>\n",
       "      <td>Homo sapiens cDNA FLJ51705 complete cds, highl...</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>ATGCGACCTCATCTTTGTCAGTGCACAAAATGGCGCCCTACAGCCT...</td>\n",
       "      <td>aconitate hydratase</td>\n",
       "      <td>Besides interconverting citrate and cis-aconit...</td>\n",
       "      <td>4.2.1.3</td>\n",
       "      <td>citric acid cycle</td>\n",
       "      <td>The citrate cycle (TCA cycle, Krebs cycle) is ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>AK303184.1</td>\n",
       "      <td>Homo sapiens cDNA FLJ52962 complete cds, highl...</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>AACTTGCTTTTGGGAGCCAGCGGTATGGCGTCGGGCTGCAAGATTG...</td>\n",
       "      <td>ribulose-phosphate 3-epimerase</td>\n",
       "      <td>The enzyme also converts D-erythrose 4-phospha...</td>\n",
       "      <td>5.1.3.1</td>\n",
       "      <td>pentose phosphate pathway</td>\n",
       "      <td>The pentose phosphate pathway is a process of ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>NM_138801.3</td>\n",
       "      <td>Homo sapiens galactose mutarotase (GALM), mRNA</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>GCTGGAGTTTGAAGAGCGGGCAGTGGCTGCACACGCCAAACTTTCC...</td>\n",
       "      <td>aldose 1-epimerase</td>\n",
       "      <td>This enzyme catalyses the first step in galact...</td>\n",
       "      <td>5.1.3.3</td>\n",
       "      <td>glycolysis</td>\n",
       "      <td>Glycolysis is the process of converting glucos...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>AK294396.1</td>\n",
       "      <td>Homo sapiens cDNA FLJ56429 complete cds, highl...</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>ACACATATGTATCCACAGCGCCTAGAACACTGCCTGTAACGTGGAA...</td>\n",
       "      <td>glucose-6-phosphate isomerase</td>\n",
       "      <td>Also catalyses the anomerization of D-glucose ...</td>\n",
       "      <td>5.3.1.9</td>\n",
       "      <td>pentose phosphate pathway</td>\n",
       "      <td>The pentose phosphate pathway is a process of ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>AK300100.1</td>\n",
       "      <td>Homo sapiens cDNA FLJ50663 complete cds, highl...</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>GACCTGCTTTGTGTGTGGGTGCGAGTGGGTGTGTGGCCCTTAACTT...</td>\n",
       "      <td>phosphoglucomutase</td>\n",
       "      <td>Maximum activity is only obtained in the prese...</td>\n",
       "      <td>5.4.2.2</td>\n",
       "      <td>glycolysis</td>\n",
       "      <td>Glycolysis is the process of converting glucos...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>NM_001298527.1</td>\n",
       "      <td>Drosophila melanogaster zwischenferment, trans...</td>\n",
       "      <td>Drosophila melanogaster</td>\n",
       "      <td>AGTTTTGACAGTTGAAAAAAGTTACTTAGTGAAAAAAACAAGTAGA...</td>\n",
       "      <td>glucose-6-phosphate dehydrogenase (NADP+)</td>\n",
       "      <td>The enzyme catalyses a step of the pentose pho...</td>\n",
       "      <td>1.1.1.49</td>\n",
       "      <td>pentose phosphate pathway</td>\n",
       "      <td>The pentose phosphate pathway is a process of ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>NM_167019.2</td>\n",
       "      <td>Drosophila melanogaster pyruvate dehydrogenase...</td>\n",
       "      <td>Drosophila melanogaster</td>\n",
       "      <td>CAAACGGTTGTTATAGTTGTGCAATATAGTAGTTGATAAACTAATC...</td>\n",
       "      <td>pyruvate dehydrogenase (acetyl-transferring)</td>\n",
       "      <td>Contains thiamine diphosphate.</td>\n",
       "      <td>1.2.4.1</td>\n",
       "      <td>citric acid cycle</td>\n",
       "      <td>The citrate cycle (TCA cycle, Krebs cycle) is ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>NM_001273267.1</td>\n",
       "      <td>Drosophila melanogaster midline uncoordinated ...</td>\n",
       "      <td>Drosophila melanogaster</td>\n",
       "      <td>CGTTCCGTCTGCGTTCACACTGACCTCCGTAAGGTTATCGGCATAA...</td>\n",
       "      <td>dihydrolipoyllysine-residue acetyltransferase</td>\n",
       "      <td>Belongs to class: Transferring groups other th...</td>\n",
       "      <td>2.3.1.12</td>\n",
       "      <td>citric acid cycle</td>\n",
       "      <td>The citrate cycle (TCA cycle, Krebs cycle) is ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>NM_079935.4</td>\n",
       "      <td>Drosophila melanogaster hexokinase C (Hex-C), ...</td>\n",
       "      <td>Drosophila melanogaster</td>\n",
       "      <td>CCTGGTTAGTCGTTACTCATCGTTCGAGGCGGTTACAACTAAGACA...</td>\n",
       "      <td>hexokinase</td>\n",
       "      <td>D-Glucose, D-mannose, D-fructose, sorbitol and...</td>\n",
       "      <td>2.7.1.1</td>\n",
       "      <td>glycolysis</td>\n",
       "      <td>Glycolysis is the process of converting glucos...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>NM_132428.3</td>\n",
       "      <td>Drosophila melanogaster uncharacterized protei...</td>\n",
       "      <td>Drosophila melanogaster</td>\n",
       "      <td>GAGGGAATGGTGATTACTGACAAATCGAAGTCGAGTCGAGTTTTAT...</td>\n",
       "      <td>6-phosphogluconolactonase</td>\n",
       "      <td>Belongs to the class: Carboxylic-ester hydrolases</td>\n",
       "      <td>3.1.1.31</td>\n",
       "      <td>pentose phosphate pathway</td>\n",
       "      <td>The pentose phosphate pathway is a process of ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>NM_001103593.2</td>\n",
       "      <td>Drosophila melanogaster Glucose-6-Phosphatase ...</td>\n",
       "      <td>Drosophila melanogaster</td>\n",
       "      <td>ATTTCATATTGGCAGAAAGACTTGAAAATATAAAAGGAGCCAGTTG...</td>\n",
       "      <td>glucose-6-phosphatase</td>\n",
       "      <td>Wide distribution in animal tissues.</td>\n",
       "      <td>3.1.3.9</td>\n",
       "      <td>glycolysis</td>\n",
       "      <td>Glycolysis is the process of converting glucos...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>AA567790.1</td>\n",
       "      <td>HL01774.5prime HL Drosophila melanogaster head...</td>\n",
       "      <td>Drosophila melanogaster</td>\n",
       "      <td>AAATCGGGCACAAGGAAAGATCAAGTGAAAAGCCTACTTTCAGTGG...</td>\n",
       "      <td>phosphoenolpyruvate carboxykinase (GTP)</td>\n",
       "      <td>ITP can act as phosphate donor.</td>\n",
       "      <td>4.1.1.32</td>\n",
       "      <td>citric acid cycle</td>\n",
       "      <td>The citrate cycle (TCA cycle, Krebs cycle) is ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>NM_001299190.1</td>\n",
       "      <td>Drosophila melanogaster mitochondrial aconitas...</td>\n",
       "      <td>Drosophila melanogaster</td>\n",
       "      <td>AATTTTGGCGGGAAACAAGCCGGCGGAACATTGCTGACGCGTATGC...</td>\n",
       "      <td>aconitate hydratase</td>\n",
       "      <td>Besides interconverting citrate and cis-aconit...</td>\n",
       "      <td>4.2.1.3</td>\n",
       "      <td>citric acid cycle</td>\n",
       "      <td>The citrate cycle (TCA cycle, Krebs cycle) is ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>NM_001273851.1</td>\n",
       "      <td>Drosophila melanogaster Ribulose-5-phosphate-3...</td>\n",
       "      <td>Drosophila melanogaster</td>\n",
       "      <td>AGCTCCCCTCTGTGTGACCTTAAGTCAACAAGACGCCACCCGCGAA...</td>\n",
       "      <td>ribulose-phosphate 3-epimerase</td>\n",
       "      <td>The enzyme also converts D-erythrose 4-phospha...</td>\n",
       "      <td>5.1.3.1</td>\n",
       "      <td>pentose phosphate pathway</td>\n",
       "      <td>The pentose phosphate pathway is a process of ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>NM_135670.2</td>\n",
       "      <td>Drosophila melanogaster uncharacterized protei...</td>\n",
       "      <td>Drosophila melanogaster</td>\n",
       "      <td>ATTTTAAATCAAAATAACCTGGCATTTTCTATCAATTTTTAATGAG...</td>\n",
       "      <td>aldose 1-epimerase</td>\n",
       "      <td>This enzyme catalyses the first step in galact...</td>\n",
       "      <td>5.1.3.3</td>\n",
       "      <td>glycolysis</td>\n",
       "      <td>Glycolysis is the process of converting glucos...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>NM_165636.2</td>\n",
       "      <td>Drosophila melanogaster phosphoglucose isomera...</td>\n",
       "      <td>Drosophila melanogaster</td>\n",
       "      <td>ATTTGACATTGAAAGGTGCTCGCGAATTCCAGAAAGGACGTTGGAA...</td>\n",
       "      <td>glucose-6-phosphate isomerase</td>\n",
       "      <td>Also catalyses the anomerization of D-glucose ...</td>\n",
       "      <td>5.3.1.9</td>\n",
       "      <td>pentose phosphate pathway</td>\n",
       "      <td>The pentose phosphate pathway is a process of ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>NM_079936.3</td>\n",
       "      <td>Drosophila melanogaster phosphoglucose mutase ...</td>\n",
       "      <td>Drosophila melanogaster</td>\n",
       "      <td>TCCAGTTGGAATTCAGCCAGCAGCCGGAAAACTCCAGTGACATGTC...</td>\n",
       "      <td>phosphoglucomutase</td>\n",
       "      <td>Maximum activity is only obtained in the prese...</td>\n",
       "      <td>5.4.2.2</td>\n",
       "      <td>glycolysis</td>\n",
       "      <td>Glycolysis is the process of converting glucos...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              gene                                   gene_description  \\\n",
       "0       AK302341.1  Homo sapiens cDNA FLJ56794 complete cds, highl...   \n",
       "1       AK296457.1  Homo sapiens cDNA FLJ52314 complete cds, highl...   \n",
       "2       AK299562.1  Homo sapiens cDNA FLJ51063 complete cds, highl...   \n",
       "3       AK294428.1  Homo sapiens cDNA FLJ56506 complete cds, highl...   \n",
       "4       CD580465.2  EST_PSF016769 hEx1 (RZPD no. 800) Homo sapiens...   \n",
       "5       AK303771.1  Homo sapiens cDNA FLJ51443 complete cds, highl...   \n",
       "6       AK316206.1  Homo sapiens cDNA, FLJ79105 complete cds, high...   \n",
       "7       AK302691.1  Homo sapiens cDNA FLJ51705 complete cds, highl...   \n",
       "8       AK303184.1  Homo sapiens cDNA FLJ52962 complete cds, highl...   \n",
       "9      NM_138801.3     Homo sapiens galactose mutarotase (GALM), mRNA   \n",
       "10      AK294396.1  Homo sapiens cDNA FLJ56429 complete cds, highl...   \n",
       "11      AK300100.1  Homo sapiens cDNA FLJ50663 complete cds, highl...   \n",
       "12  NM_001298527.1  Drosophila melanogaster zwischenferment, trans...   \n",
       "13     NM_167019.2  Drosophila melanogaster pyruvate dehydrogenase...   \n",
       "14  NM_001273267.1  Drosophila melanogaster midline uncoordinated ...   \n",
       "15     NM_079935.4  Drosophila melanogaster hexokinase C (Hex-C), ...   \n",
       "16     NM_132428.3  Drosophila melanogaster uncharacterized protei...   \n",
       "17  NM_001103593.2  Drosophila melanogaster Glucose-6-Phosphatase ...   \n",
       "18      AA567790.1  HL01774.5prime HL Drosophila melanogaster head...   \n",
       "19  NM_001299190.1  Drosophila melanogaster mitochondrial aconitas...   \n",
       "20  NM_001273851.1  Drosophila melanogaster Ribulose-5-phosphate-3...   \n",
       "21     NM_135670.2  Drosophila melanogaster uncharacterized protei...   \n",
       "22     NM_165636.2  Drosophila melanogaster phosphoglucose isomera...   \n",
       "23     NM_079936.3  Drosophila melanogaster phosphoglucose mutase ...   \n",
       "\n",
       "                   organism  \\\n",
       "0              Homo sapiens   \n",
       "1              Homo sapiens   \n",
       "2              Homo sapiens   \n",
       "3              Homo sapiens   \n",
       "4              Homo sapiens   \n",
       "5              Homo sapiens   \n",
       "6              Homo sapiens   \n",
       "7              Homo sapiens   \n",
       "8              Homo sapiens   \n",
       "9              Homo sapiens   \n",
       "10             Homo sapiens   \n",
       "11             Homo sapiens   \n",
       "12  Drosophila melanogaster   \n",
       "13  Drosophila melanogaster   \n",
       "14  Drosophila melanogaster   \n",
       "15  Drosophila melanogaster   \n",
       "16  Drosophila melanogaster   \n",
       "17  Drosophila melanogaster   \n",
       "18  Drosophila melanogaster   \n",
       "19  Drosophila melanogaster   \n",
       "20  Drosophila melanogaster   \n",
       "21  Drosophila melanogaster   \n",
       "22  Drosophila melanogaster   \n",
       "23  Drosophila melanogaster   \n",
       "\n",
       "                                              nuc_seq  \\\n",
       "0   AAAACGGTCGTACACTTCGGGGCTGCGAGCGCGGAGGGCGACGACG...   \n",
       "1   ACCCGCGTCGTGCCTCCTGGGTTGTGAGGAGTCGCCGCTGCCGCCA...   \n",
       "2   CCCGTTCGTCGCAACAGCGTGACTACAGGGTATGGCGGGGTCCGGG...   \n",
       "3   GAGGAGCCGCCGAGCAGCCGCCGGAGGACCACGGCTCGCCAGGGCT...   \n",
       "4   CTCCCCGCCGCCGCCCTCGCCATGGCCGCGCCGGCCCCGGGCCTCA...   \n",
       "5   ATAGCAGAGCAATCACCACCAAGCCTGGAATAACTGCAAGGGCTCT...   \n",
       "6   GTGGCTCGCTTCGCCGCGCTCCCTCCTTCCCCGCCTTCCATACCTC...   \n",
       "7   ATGCGACCTCATCTTTGTCAGTGCACAAAATGGCGCCCTACAGCCT...   \n",
       "8   AACTTGCTTTTGGGAGCCAGCGGTATGGCGTCGGGCTGCAAGATTG...   \n",
       "9   GCTGGAGTTTGAAGAGCGGGCAGTGGCTGCACACGCCAAACTTTCC...   \n",
       "10  ACACATATGTATCCACAGCGCCTAGAACACTGCCTGTAACGTGGAA...   \n",
       "11  GACCTGCTTTGTGTGTGGGTGCGAGTGGGTGTGTGGCCCTTAACTT...   \n",
       "12  AGTTTTGACAGTTGAAAAAAGTTACTTAGTGAAAAAAACAAGTAGA...   \n",
       "13  CAAACGGTTGTTATAGTTGTGCAATATAGTAGTTGATAAACTAATC...   \n",
       "14  CGTTCCGTCTGCGTTCACACTGACCTCCGTAAGGTTATCGGCATAA...   \n",
       "15  CCTGGTTAGTCGTTACTCATCGTTCGAGGCGGTTACAACTAAGACA...   \n",
       "16  GAGGGAATGGTGATTACTGACAAATCGAAGTCGAGTCGAGTTTTAT...   \n",
       "17  ATTTCATATTGGCAGAAAGACTTGAAAATATAAAAGGAGCCAGTTG...   \n",
       "18  AAATCGGGCACAAGGAAAGATCAAGTGAAAAGCCTACTTTCAGTGG...   \n",
       "19  AATTTTGGCGGGAAACAAGCCGGCGGAACATTGCTGACGCGTATGC...   \n",
       "20  AGCTCCCCTCTGTGTGACCTTAAGTCAACAAGACGCCACCCGCGAA...   \n",
       "21  ATTTTAAATCAAAATAACCTGGCATTTTCTATCAATTTTTAATGAG...   \n",
       "22  ATTTGACATTGAAAGGTGCTCGCGAATTCCAGAAAGGACGTTGGAA...   \n",
       "23  TCCAGTTGGAATTCAGCCAGCAGCCGGAAAACTCCAGTGACATGTC...   \n",
       "\n",
       "                                           enzyme  \\\n",
       "0       glucose-6-phosphate dehydrogenase (NADP+)   \n",
       "1    pyruvate dehydrogenase (acetyl-transferring)   \n",
       "2   dihydrolipoyllysine-residue acetyltransferase   \n",
       "3                                      hexokinase   \n",
       "4                       6-phosphogluconolactonase   \n",
       "5                           glucose-6-phosphatase   \n",
       "6         phosphoenolpyruvate carboxykinase (GTP)   \n",
       "7                             aconitate hydratase   \n",
       "8                  ribulose-phosphate 3-epimerase   \n",
       "9                              aldose 1-epimerase   \n",
       "10                  glucose-6-phosphate isomerase   \n",
       "11                             phosphoglucomutase   \n",
       "12      glucose-6-phosphate dehydrogenase (NADP+)   \n",
       "13   pyruvate dehydrogenase (acetyl-transferring)   \n",
       "14  dihydrolipoyllysine-residue acetyltransferase   \n",
       "15                                     hexokinase   \n",
       "16                      6-phosphogluconolactonase   \n",
       "17                          glucose-6-phosphatase   \n",
       "18        phosphoenolpyruvate carboxykinase (GTP)   \n",
       "19                            aconitate hydratase   \n",
       "20                 ribulose-phosphate 3-epimerase   \n",
       "21                             aldose 1-epimerase   \n",
       "22                  glucose-6-phosphate isomerase   \n",
       "23                             phosphoglucomutase   \n",
       "\n",
       "                                   enzyme_description     ecnum  \\\n",
       "0   The enzyme catalyses a step of the pentose pho...  1.1.1.49   \n",
       "1                      Contains thiamine diphosphate.   1.2.4.1   \n",
       "2   Belongs to class: Transferring groups other th...  2.3.1.12   \n",
       "3   D-Glucose, D-mannose, D-fructose, sorbitol and...   2.7.1.1   \n",
       "4   Belongs to the class: Carboxylic-ester hydrolases  3.1.1.31   \n",
       "5                Wide distribution in animal tissues.   3.1.3.9   \n",
       "6                     ITP can act as phosphate donor.  4.1.1.32   \n",
       "7   Besides interconverting citrate and cis-aconit...   4.2.1.3   \n",
       "8   The enzyme also converts D-erythrose 4-phospha...   5.1.3.1   \n",
       "9   This enzyme catalyses the first step in galact...   5.1.3.3   \n",
       "10  Also catalyses the anomerization of D-glucose ...   5.3.1.9   \n",
       "11  Maximum activity is only obtained in the prese...   5.4.2.2   \n",
       "12  The enzyme catalyses a step of the pentose pho...  1.1.1.49   \n",
       "13                     Contains thiamine diphosphate.   1.2.4.1   \n",
       "14  Belongs to class: Transferring groups other th...  2.3.1.12   \n",
       "15  D-Glucose, D-mannose, D-fructose, sorbitol and...   2.7.1.1   \n",
       "16  Belongs to the class: Carboxylic-ester hydrolases  3.1.1.31   \n",
       "17               Wide distribution in animal tissues.   3.1.3.9   \n",
       "18                    ITP can act as phosphate donor.  4.1.1.32   \n",
       "19  Besides interconverting citrate and cis-aconit...   4.2.1.3   \n",
       "20  The enzyme also converts D-erythrose 4-phospha...   5.1.3.1   \n",
       "21  This enzyme catalyses the first step in galact...   5.1.3.3   \n",
       "22  Also catalyses the anomerization of D-glucose ...   5.3.1.9   \n",
       "23  Maximum activity is only obtained in the prese...   5.4.2.2   \n",
       "\n",
       "                      pathway  \\\n",
       "0   pentose phosphate pathway   \n",
       "1           citric acid cycle   \n",
       "2           citric acid cycle   \n",
       "3                  glycolysis   \n",
       "4   pentose phosphate pathway   \n",
       "5                  glycolysis   \n",
       "6           citric acid cycle   \n",
       "7           citric acid cycle   \n",
       "8   pentose phosphate pathway   \n",
       "9                  glycolysis   \n",
       "10  pentose phosphate pathway   \n",
       "11                 glycolysis   \n",
       "12  pentose phosphate pathway   \n",
       "13          citric acid cycle   \n",
       "14          citric acid cycle   \n",
       "15                 glycolysis   \n",
       "16  pentose phosphate pathway   \n",
       "17                 glycolysis   \n",
       "18          citric acid cycle   \n",
       "19          citric acid cycle   \n",
       "20  pentose phosphate pathway   \n",
       "21                 glycolysis   \n",
       "22  pentose phosphate pathway   \n",
       "23                 glycolysis   \n",
       "\n",
       "                                  pathway_description  \n",
       "0   The pentose phosphate pathway is a process of ...  \n",
       "1   The citrate cycle (TCA cycle, Krebs cycle) is ...  \n",
       "2   The citrate cycle (TCA cycle, Krebs cycle) is ...  \n",
       "3   Glycolysis is the process of converting glucos...  \n",
       "4   The pentose phosphate pathway is a process of ...  \n",
       "5   Glycolysis is the process of converting glucos...  \n",
       "6   The citrate cycle (TCA cycle, Krebs cycle) is ...  \n",
       "7   The citrate cycle (TCA cycle, Krebs cycle) is ...  \n",
       "8   The pentose phosphate pathway is a process of ...  \n",
       "9   Glycolysis is the process of converting glucos...  \n",
       "10  The pentose phosphate pathway is a process of ...  \n",
       "11  Glycolysis is the process of converting glucos...  \n",
       "12  The pentose phosphate pathway is a process of ...  \n",
       "13  The citrate cycle (TCA cycle, Krebs cycle) is ...  \n",
       "14  The citrate cycle (TCA cycle, Krebs cycle) is ...  \n",
       "15  Glycolysis is the process of converting glucos...  \n",
       "16  The pentose phosphate pathway is a process of ...  \n",
       "17  Glycolysis is the process of converting glucos...  \n",
       "18  The citrate cycle (TCA cycle, Krebs cycle) is ...  \n",
       "19  The citrate cycle (TCA cycle, Krebs cycle) is ...  \n",
       "20  The pentose phosphate pathway is a process of ...  \n",
       "21  Glycolysis is the process of converting glucos...  \n",
       "22  The pentose phosphate pathway is a process of ...  \n",
       "23  Glycolysis is the process of converting glucos...  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "conn = sqlite3.connect('lab3.db')\n",
    "select_command = \"\"\"\n",
    "    SELECT genes.name AS gene, genes.description AS gene_description, organism, nuc_seq, \n",
    "    enzymes.name AS enzyme, enzymes.description AS enzyme_description, ecnum, \n",
    "    pathway, pathways.description AS pathway_description  \n",
    "    FROM genes\n",
    "    LEFT JOIN enzymes ON genes.enzyme_ecnum = enzymes.ecnum\n",
    "    LEFT JOIN pathways ON enzymes.pathway = pathways.name\n",
    "\"\"\"\n",
    "\n",
    "pd.read_sql(select_command, conn)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Part 6.2\n",
    "Retrieve just the description of all the genes associated with glycolisis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Homo sapiens cDNA FLJ56506 complete cds, highl...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Homo sapiens cDNA FLJ51443 complete cds, highl...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Homo sapiens galactose mutarotase (GALM), mRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Homo sapiens cDNA FLJ50663 complete cds, highl...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Drosophila melanogaster hexokinase C (Hex-C), ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Drosophila melanogaster Glucose-6-Phosphatase ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Drosophila melanogaster uncharacterized protei...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Drosophila melanogaster phosphoglucose mutase ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    gene_description\n",
       "0  Homo sapiens cDNA FLJ56506 complete cds, highl...\n",
       "1  Homo sapiens cDNA FLJ51443 complete cds, highl...\n",
       "2     Homo sapiens galactose mutarotase (GALM), mRNA\n",
       "3  Homo sapiens cDNA FLJ50663 complete cds, highl...\n",
       "4  Drosophila melanogaster hexokinase C (Hex-C), ...\n",
       "5  Drosophila melanogaster Glucose-6-Phosphatase ...\n",
       "6  Drosophila melanogaster uncharacterized protei...\n",
       "7  Drosophila melanogaster phosphoglucose mutase ..."
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "conn = sqlite3.connect('lab3.db')\n",
    "select_command = \"\"\"\n",
    "    SELECT genes.description AS gene_description \n",
    "    FROM genes\n",
    "    LEFT JOIN enzymes ON genes.enzyme_ecnum = enzymes.ecnum\n",
    "    LEFT JOIN pathways ON enzymes.pathway = pathways.name\n",
    "    WHERE pathway = 'glycolysis'\n",
    "\"\"\"\n",
    "\n",
    "pd.read_sql(select_command, conn)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Part 6.3\n",
    "Are the joins one-to-one, many-to-one, or one-to-many?\n",
    "\n",
    "### Answer"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The joins are many-to-one.\n",
    "\n",
    "We start with the genes tables as the 'backbone' with 24 rows, then execute a 2-to-1 join to bring in the genes data, followed by an 8-to-1 join to bring in the pathways data."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Submission\n",
    "\n",
    "1. Commit and push this notebook,\n",
    "2. Git add and commit the database file (then push)\n",
    "3. Submit repo link on GitHub"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
